File: kernel/system/time/clock.c
// Clock registration.

#include <menix/common.h>
#include <menix/system/arch.h>
#include <menix/system/time/clock.h>
#include <menix/util/log.h>

static ClockSource* current_source = NULL;
static usize clock_base_ns = 0;	   // Base counter in nanoseconds.

void clock_register(ClockSource* source)
{
	if (source == NULL || source->get_elapsed_ns == NULL || source->reset == NULL)
		return;

	print_log("clock: Switching to new source \"%s\"\n", source->name);

	// Move the current value over to the new counter.
	const usize timer = clock_get_elapsed_ns();
	current_source = source;
	clock_set_elapsed_ns(timer);
}

usize clock_get_elapsed_ns()
{
	if (current_source != NULL)
		return current_source->get_elapsed_ns();

	return 0;
}

void clock_set_elapsed_ns(usize value)
{
	clock_base_ns = value;
	if (current_source != NULL)
		current_source->reset();
}

void clock_wait(usize ns)
{
	if (current_source == NULL)
	{
		print_warn("clock: Attempted to wait %zu nanoseconds, "
				   "but this would hang indefinitely since no clock source is available.\n",
				   ns);
		return;
	}

	usize time = clock_get_elapsed_ns() + ns;
	while (time > clock_get_elapsed_ns())
	{
		asm_pause();
	}
}

File: kernel/src/generic/clock.rs
//! Global timer management.
// TODO: Try to get rid of some locks.

use super::util::spin_mutex::SpinMutex;
use alloc::boxed::Box;

#[initgraph::task(name = "generic.clock")]
pub fn CLOCK_STAGE() {}

pub trait ClockSource: Send {
    fn name(&self) -> &'static str;

    /// A priority of a clock source. A high value equals a high priority.
    fn get_priority(&self) -> u8;

    /// Sets the elapsed nanoseconds to start counting at.
    fn reset(&mut self);

    /// Gets the elapsed nanoseconds since initialization of this timer.
    fn get_elapsed_ns(&self) -> usize;
}

#[derive(Debug, PartialEq)]
pub enum ClockError {
    /// The clock source has a lesser priority.
    LowerPriority,
    /// The clock source is unavailable.
    Unavailable,
    /// The clock source is not sane.
    InvalidConfiguration,
    /// The clock source could not be calibrated.
    UnableToSetup,
}

/// Gets the elapsed nanoseconds since initialization of this timer.
pub fn get_elapsed() -> usize {
    let guard = CLOCK.lock();
    match &guard.current {
        Some(x) => x.get_elapsed_ns() + guard.counter_base,
        None => 0,
    }
}

/// Switches to a new clock source if it is of higher priority.
pub fn switch(mut new_source: Box<dyn ClockSource>) -> Result<(), ClockError> {
    // Determine if we should make the switch.
    if let Some(x) = &CLOCK.lock().current {
        let prio = x.get_priority();
        if new_source.get_priority() > prio {
            Ok(())
        } else {
            Err(ClockError::LowerPriority)
        }
    } else {
        Ok(())
    }?;

    log!("Switching to clock source \"{}\"", new_source.name());

    // Save the current counter.
    let elapsed = get_elapsed();
    let mut clock = CLOCK.lock();
    clock.counter_base = elapsed;

    new_source.reset();
    clock.current = Some(new_source);
    return Ok(());
}

pub fn has_clock() -> bool {
    return CLOCK.lock().current.is_some();
}

/// Blocking wait for a given amount of nanoseconds.
pub fn block_ns(time: usize) -> Result<(), ClockError> {
    if CLOCK.lock().current.is_none() {
        error!(
            "Unable to sleep for {} nanoseconds. No clock source available, this would block forever!",
            time
        );
        return Err(ClockError::Unavailable);
    }

    let target = get_elapsed() + time;
    while get_elapsed() < target {}
    return Ok(());
}

struct Clock {
    /// The active clock source.
    current: Option<Box<dyn ClockSource>>,
    /// An offset to add to the read counter.
    counter_base: usize,
}

static CLOCK: SpinMutex<Clock> = SpinMutex::new(Clock {
    current: None,
    counter_base: 0,
});
